File: shopping-cart/typescript/src/Product.ts
import type { DiscountPolicy } from './DiscountPolicy.js';
import { NoDiscount } from './DiscountPolicy.js';
import { Money } from './Money.js';

export class Product {
  readonly discountPolicy: DiscountPolicy;

  constructor(
    public readonly sku: string,
    public readonly name: string,
    public readonly unitPrice: Money,
    discountPolicy?: DiscountPolicy,
  ) {
    if (sku.trim().length === 0) {
      throw new Error('SKU must not be empty');
    }
    this.discountPolicy = discountPolicy ?? NoDiscount.instance;
  }
}

File: shopping-cart/csharp/src/ShoppingCart/Product.cs
namespace ShoppingCart;

public sealed class Product
{
    public string Sku { get; }
    public string Name { get; }
    public Money UnitPrice { get; }
    public IDiscountPolicy DiscountPolicy { get; }

    public Product(string sku, string name, Money unitPrice, IDiscountPolicy? discountPolicy = null)
    {
        if (string.IsNullOrWhiteSpace(sku))
        {
            throw new ArgumentException("SKU must not be empty", nameof(sku));
        }
        Sku = sku;
        Name = name;
        UnitPrice = unitPrice;
        DiscountPolicy = discountPolicy ?? NoDiscount.Instance;
    }
}
